File: Cpp/Legacy89DiskKit.Cpp/src/infrastructure/filesystem/n88_basic/n88_basic_file_system.cpp
#include "legacy89diskkit/cpp/infrastructure/filesystem/n88_basic/n88_basic_file_system.hpp"

#include "legacy89diskkit/cpp/hu_basic_name_rules.hpp"
#include "legacy89diskkit/cpp/n88_basic_allocation_rules.hpp"
#include "legacy89diskkit/cpp/n88_basic_attribute_update_transaction.hpp"
#include "legacy89diskkit/cpp/n88_basic_configuration.hpp"
#include "legacy89diskkit/cpp/n88_basic_delete_transaction.hpp"
#include "legacy89diskkit/cpp/n88_basic_dir_parser.hpp"
#include "legacy89diskkit/cpp/n88_basic_fat_rules.hpp"
#include "legacy89diskkit/cpp/n88_basic_format_rules.hpp"
#include "legacy89diskkit/cpp/n88_basic_read_rules.hpp"
#include "legacy89diskkit/cpp/n88_basic_rename_transaction.hpp"
#include "legacy89diskkit/cpp/n88_basic_shell.hpp"
#include "legacy89diskkit/cpp/n88_basic_write_transaction.hpp"

#include <algorithm>
#include <array>
#include <string>

namespace legacy89diskkit::cpp
{
namespace
{
std::string NormalizeName(const std::string_view name)
{
    std::string normalized(name);
    std::transform(
        normalized.begin(),
        normalized.end(),
        normalized.begin(),
        [](const unsigned char ch)
        {
            return static_cast<char>(std::toupper(ch));
        });
    return normalized;
}

std::string BuildFullName(const N88BasicFileEntry& entry)
{
    return HuBasicNameRules::BuildDisplayName(entry.file_name, entry.extension);
}

std::array<std::uint8_t, 16> EncodeDirectoryEntry(const N88BasicFileEntry& entry)
{
    auto encoded = N88BasicDirParser::Write(entry);
    for (int index = 0; index < 6; ++index)
    {
        if (encoded[index] == 0x00)
        {
            encoded[index] = static_cast<std::uint8_t>(' ');
        }
    }

    for (int index = 6; index < 9; ++index)
    {
        if (encoded[index] == 0x00)
        {
            encoded[index] = static_cast<std::uint8_t>(' ');
        }
    }

    return encoded;
}
}

N88BasicFileSystem N88BasicFileSystem::Open(RawDiskContainer& container)
{
    return N88BasicFileSystem(&container, nullptr, container.DiskTypeValue(), container.IsReadOnly());
}

N88BasicFileSystem N88BasicFileSystem::Open(D88DiskContainer& container)
{
    return N88BasicFileSystem(nullptr, &container, container.DiskTypeValue(), container.IsReadOnly());
}

N88BasicFileSystem::N88BasicFileSystem(
    RawDiskContainer* raw_container,
    D88DiskContainer* d88_container,
    const DiskType disk_type,
    const bool read_only)
    : raw_container_(raw_container),
      d88_container_(d88_container),
      config_(N88BasicConfigurationProvider::GetDefault(disk_type)),
      disk_type_(disk_type),
      read_only_(read_only)
{
}

const N88BasicConfiguration& N88BasicFileSystem::GetConfiguration() const
{
    return config_;
}

DiskType N88BasicFileSystem::DiskTypeValue() const
{
    return disk_type_;
}

bool N88BasicFileSystem::IsReadOnly() const
{
    return read_only_;
}

N88BasicFileSystemInfo N88BasicFileSystem::GetFileSystemInfo() const
{
    return N88BasicShell::GetFileSystemInfo(ReadFat(), config_);
}

std::vector<N88BasicFileEntry> N88BasicFileSystem::GetFiles() const
{
    return N88BasicShell::ListFiles(ReadDirectorySectors(), ReadFat(), config_);
}

bool N88BasicFileSystem::FileExists(const std::string_view file_name) const
{
    const auto file_name_text = std::string(file_name);
    return N88BasicShell::FileExists(ReadDirectorySectors(), ReadFat(), config_, file_name_text.c_str());
}

Result<std::vector<std::uint8_t>> N88BasicFileSystem::ReadFile(const std::string_view file_name) const
{
    const auto target = NormalizeName(file_name);
    const auto files = GetFiles();
    const auto it = std::find_if(
        files.begin(),
        files.end(),
        [&](const auto& entry)
        {
            return NormalizeName(BuildFullName(entry)) == target;
        });
    if (it == files.end())
    {
        return Result<std::vector<std::uint8_t>>::Failure(StatusCode::InvalidArgument, "File not found.");
    }

    const auto fat_data = ReadFat();
    const auto chain = N88BasicFatRules::GetClusterChain(fat_data, config_, it->start_cluster);
    std::vector<std::uint8_t> data;
    for (const auto cluster : chain)
    {
        const auto cluster_data = ReadCluster(cluster);
        if (!cluster_data.ok())
        {
            return cluster_data;
        }

        data.insert(data.end(), cluster_data.value().begin(), cluster_data.value().end());
    }

    auto file = *it;
    file.size = N88BasicReadRules::ResolveSizeFromFat(chain, fat_data, config_);
    return Result<std::vector<std::uint8_t>>::Success(N88BasicReadRules::ResolveReadPayload(data, file));
}

Status N88BasicFileSystem::WriteFile(
    const std::string_view file_name,
    const std::vector<std::uint8_t>& data,
    const N88BasicFileAttributes& attributes)
{
    if (read_only_)
    {
        return {StatusCode::InvalidArgument, "Filesystem is read-only."};
    }

    if (FileExists(file_name))
    {
        return {StatusCode::InvalidArgument, "File already exists."};
    }

    auto fat_data = ReadFat();
    const auto file_name_text = std::string(file_name);
    const auto plan = N88BasicShell::PlanWrite(file_name_text.c_str(), data, attributes, config_, fat_data);
    if (!plan.has_value())
    {
        return {StatusCode::InvalidArgument, "Unable to create write plan."};
    }

    const auto cluster_size = static_cast<std::size_t>(config_.cluster_size);
    for (std::size_t index = 0; index < plan->allocated_clusters.size(); ++index)
    {
        std::vector<std::uint8_t> cluster_buffer(cluster_size, 0x00);
        const auto offset = index * cluster_size;
        const auto count = std::min(cluster_size, plan->payload.size() - std::min(plan->payload.size(), offset));
        if (count > 0)
        {
            std::copy_n(plan->payload.begin() + static_cast<std::ptrdiff_t>(offset), count, cluster_buffer.begin());
        }

        const auto status = WriteCluster(plan->allocated_clusters[index], cluster_buffer);
        if (!status.ok())
        {
            return status;
        }
    }

    for (std::size_t index = 0; index < plan->allocated_clusters.size(); ++index)
    {
        const auto next = index + 1 == plan->allocated_clusters.size()
            ? plan->terminal_flag
            : plan->allocated_clusters[index + 1];
        N88BasicFatRules::SetEntry(fat_data, plan->allocated_clusters[index], next);
    }

    const auto fat_status = WriteFat(fat_data);
    if (!fat_status.ok())
    {
        return fat_status;
    }

    auto directory_sectors = ReadDirectorySectors();
    for (std::size_t sector_index = 0; sector_index < directory_sectors.size(); ++sector_index)
    {
        for (auto offset = 0; offset < config_.sector_size; offset += 16)
        {
            const auto marker = directory_sectors[sector_index][offset];
            if (marker != 0x00 && marker != 0xff)
            {
                continue;
            }

            const auto encoded = EncodeDirectoryEntry(plan->file_entry);
            std::copy(encoded.begin(), encoded.end(), directory_sectors[sector_index].begin() + offset);
            if (marker == 0xff && offset + 16 < config_.sector_size)
            {
                directory_sectors[sector_index][offset + 16] = 0xff;
            }

            return WriteDirectorySector(static_cast<int>(sector_index), directory_sectors[sector_index]);
        }
    }

    return {StatusCode::OutOfRange, "Directory is full."};
}

Status N88BasicFileSystem::DeleteFile(const std::string_view file_name)
{
    if (read_only_)
    {
        return {StatusCode::InvalidArgument, "Filesystem is read-only."};
    }

    const auto directory_sectors = ReadDirectorySectors();
    const auto file_name_text = std::string(file_name);
    const auto file = N88BasicShell::FindFile(directory_sectors, ReadFat(), config_, file_name_text.c_str());
    if (!file.has_value())
    {
        return {StatusCode::InvalidArgument, "File not found."};
    }

    auto fat_data = ReadFat();
    const auto chain = N88BasicFatRules::GetClusterChain(fat_data, config_, file->start_cluster);
    const auto plan = N88BasicShell::PlanDelete(fat_data, chain, directory_sectors, config_, file_name_text.c_str());
    if (!plan.has_value())
    {
        return {StatusCode::InvalidArgument, "Unable to create delete plan."};
    }

    const auto fat_status = WriteFat(plan->fat_data);
    if (!fat_status.ok())
    {
        return fat_status;
    }

    auto sector = directory_sectors[plan->sector_index];
    sector[plan->entry_offset] = 0x00;
    return WriteDirectorySector(plan->sector_index, sector);
}

Status N88BasicFileSystem::RenameFile(const std::string_view old_name, const std::string_view new_name)
{
    if (read_only_)
    {
        return {StatusCode::InvalidArgument, "Filesystem is read-only."};
    }

    auto directory_sectors = ReadDirectorySectors();
    const auto old_name_text = std::string(old_name);
    const auto new_name_text = std::string(new_name);
    const auto plan = N88BasicShell::PlanRename(directory_sectors, config_, old_name_text.c_str(), new_name_text.c_str());
    if (!plan.has_value())
    {
        return {StatusCode::InvalidArgument, "Unable to create rename plan."};
    }

    const auto encoded = EncodeDirectoryEntry(plan->updated_entry);
    std::copy(encoded.begin(), encoded.end(), directory_sectors[plan->sector_index].begin() + plan->entry_offset);
    return WriteDirectorySector(plan->sector_index, directory_sectors[plan->sector_index]);
}

Status N88BasicFileSystem::UpdateAttributes(const std::string_view file_name, const N88BasicFileAttributes& attributes)
{
    if (read_only_)
    {
        return {StatusCode::InvalidArgument, "Filesystem is read-only."};
    }

    auto directory_sectors = ReadDirectorySectors();
    const auto file_name_text = std::string(file_name);
    const auto plan = N88BasicShell::PlanAttributeUpdate(directory_sectors, config_, file_name_text.c_str(), attributes);
    if (!plan.has_value())
    {
        return {StatusCode::InvalidArgument, "Unable to create attribute update plan."};
    }

    const auto encoded = EncodeDirectoryEntry(plan->updated_entry);
    std::copy(encoded.begin(), encoded.end(), directory_sectors[plan->sector_index].begin() + plan->entry_offset);
    return WriteDirectorySector(plan->sector_index, directory_sectors[plan->sector_index]);
}

Status N88BasicFileSystem::Format()
{
    if (read_only_)
    {
        return {StatusCode::InvalidArgument, "Filesystem is read-only."};
    }

    const auto fat_status = WriteFat(N88BasicShell::CreateFatData(config_));
    if (!fat_status.ok())
    {
        return fat_status;
    }

    const auto directory = N88BasicShell::CreateDirectorySectors(config_);
    for (std::size_t index = 0; index < directory.size(); ++index)
    {
        const auto status = WriteDirectorySector(static_cast<int>(index), directory[index]);
        if (!status.ok())
        {
            return status;
        }
    }

    return Status::OkStatus();
}

Result<std::vector<std::uint8_t>> N88BasicFileSystem::ReadBootArea() const
{
    return ReadSector(0, 0, 1);
}

Status N88BasicFileSystem::WriteBootArea(const std::vector<std::uint8_t>& data)
{
    if (read_only_)
    {
        return {StatusCode::InvalidArgument, "Filesystem is read-only."};
    }

    std::vector<std::uint8_t> sector_data(config_.sector_size, 0x00);
    const size_t copy_size = std::min<size_t>(data.size(), config_.sector_size);
    std::copy(data.begin(), data.begin() + copy_size, sector_data.begin());

    return WriteSector(0, 0, 1, sector_data);
}

Result<std::vector<std::uint8_t>> N88BasicFileSystem::ReadSector(const int cylinder, const int head, const int sector) const
{
    if (raw_container_ != nullptr)
    {
        return raw_container_->ReadSector(cylinder, head, sector);
    }

    return d88_container_->ReadSector(cylinder, head, sector);
}

Status N88BasicFileSystem::WriteSector(const int cylinder, const int head, const int sector, const std::vector<std::uint8_t>& data)
{
    if (raw_container_ != nullptr)
    {
        return raw_container_->WriteSector(cylinder, head, sector, data);
    }

    return d88_container_->WriteSector(cylinder, head, sector, data);
}

std::vector<std::uint8_t> N88BasicFileSystem::ReadFat() const
{
    std::vector<std::uint8_t> fat_data(config_.fat_sectors * config_.sector_size, 0x00);
    for (int sector_index = 0; sector_index < config_.fat_sectors; ++sector_index)
    {
        const auto sector = ReadSector(config_.system_track, config_.system_head, config_.fat_sector + sector_index);
        if (!sector.ok())
        {
            return {};
        }

        std::copy(
            sector.value().begin(),
            sector.value().begin() + config_.sector_size,
            fat_data.begin() + (sector_index * config_.sector_size));
    }

    return fat_data;
}

Status N88BasicFileSystem::WriteFat(const std::vector<std::uint8_t>& fat_data)
{
    for (int sector_index = 0; sector_index < config_.fat_sectors; ++sector_index)
    {
        std::vector<std::uint8_t> sector(config_.sector_size, 0x00);
        std::copy(
            fat_data.begin() + (sector_index * config_.sector_size),
            fat_data.begin() + ((sector_index + 1) * config_.sector_size),
            sector.begin());
        const auto status = WriteSector(config_.system_track, config_.system_head, config_.fat_sector + sector_index, sector);
        if (!status.ok())
        {
            return status;
        }
    }

    return Status::OkStatus();
}

std::vector<std::vector<std::uint8_t>> N88BasicFileSystem::ReadDirectorySectors() const
{
    std::vector<std::vector<std::uint8_t>> sectors;
    sectors.reserve(config_.directory_sectors);
    for (int sector_index = 0; sector_index < config_.directory_sectors; ++sector_index)
    {
        const auto sector = ReadSector(config_.system_track, config_.system_head, config_.directory_sector + sector_index);
        sectors.push_back(sector.ok() ? sector.value() : std::vector<std::uint8_t>(config_.sector_size, 0x00));
    }
    return sectors;
}

Status N88BasicFileSystem::WriteDirectorySector(const int sector_index, const std::vector<std::uint8_t>& sector_data)
{
    return WriteSector(config_.system_track, config_.system_head, config_.directory_sector + sector_index, sector_data);
}

Result<std::vector<std::uint8_t>> N88BasicFileSystem::ReadCluster(const int cluster) const
{
    const auto sectors_per_cluster = config_.cluster_size / config_.sector_size;
    std::vector<std::uint8_t> cluster_data(config_.cluster_size, 0x00);
    const auto start_lba = cluster * sectors_per_cluster;
    for (int sector_index = 0; sector_index < sectors_per_cluster; ++sector_index)
    {
        const auto lba = start_lba + sector_index;
        const auto track = lba / config_.sectors_per_track;
        const auto sector = (lba % config_.sectors_per_track) + 1;
        int cylinder;
        int head;
        int physical_sector;
        GetPhysicalAddress(track, sector, cylinder, head, physical_sector);
        const auto result = ReadSector(cylinder, head, physical_sector);
        if (!result.ok())
        {
            return result;
        }

        std::copy(
            result.value().begin(),
            result.value().begin() + config_.sector_size,
            cluster_data.begin() + (sector_index * config_.sector_size));
    }

    return Result<std::vector<std::uint8_t>>::Success(cluster_data);
}

Status N88BasicFileSystem::WriteCluster(const int cluster, const std::vector<std::uint8_t>& cluster_data)
{
    const auto sectors_per_cluster = config_.cluster_size / config_.sector_size;
    const auto start_lba = cluster * sectors_per_cluster;
    for (int sector_index = 0; sector_index < sectors_per_cluster; ++sector_index)
    {
        const auto lba = start_lba + sector_index;
        const auto track = lba / config_.sectors_per_track;
        const auto sector = (lba % config_.sectors_per_track) + 1;
        int cylinder;
        int head;
        int physical_sector;
        GetPhysicalAddress(track, sector, cylinder, head, physical_sector);
        std::vector<std::uint8_t> sector_data(config_.sector_size, 0x00);
        std::copy(
            cluster_data.begin() + (sector_index * config_.sector_size),
            cluster_data.begin() + ((sector_index + 1) * config_.sector_size),
            sector_data.begin());
        const auto status = WriteSector(cylinder, head, physical_sector, sector_data);
        if (!status.ok())
        {
            return status;
        }
    }

    return Status::OkStatus();
}

void N88BasicFileSystem::GetPhysicalAddress(const int track, const int sector, int& cylinder, int& head, int& physical_sector)
{
    cylinder = track / 2;
    head = track % 2;
    physical_sector = sector;
}
}

File: CSharp/Legacy89DiskKit.Infrastructure/FileSystem/Pc88/N88BasicFileSystem.cs
using Legacy89DiskKit.Domain.DiskImage.Interface.Container;
using Legacy89DiskKit.Domain.DiskImage.Model;
using Legacy89DiskKit.Domain.FileSystem.Interface.FileSystem;
using Legacy89DiskKit.Domain.FileSystem.Model;
using Legacy89DiskKit.Infrastructure.FileSystem.Pc88.Models;
using Legacy89DiskKit.Domain.FileSystem.Exception;
using DomainAttr = Legacy89DiskKit.Domain.FileSystem.Model.FileAttributes;

namespace Legacy89DiskKit.Infrastructure.FileSystem.Pc88;

public class N88BasicFileSystem : IFileSystem
{
    private readonly IDiskContainer _diskContainer;
    private readonly N88BasicConfiguration _config;
    private readonly N88BasicFatManager _fatManager;
    private readonly N88BasicDirParser _dirParser;

    public N88BasicFileSystem(IDiskContainer diskContainer)
    {
        _diskContainer = diskContainer;
        _config = N88BasicConfiguration.GetDefault(diskContainer.DiskType);
        _fatManager = new N88BasicFatManager(diskContainer, _config);
        _dirParser = new N88BasicDirParser(_config);
    }

    public DiskFileSystemInfo GetFileSystemInfo()
    {
        var fat = _fatManager.ReadFat();
        int free = 0;
        for (int i = _config.ReservedClusters; i < _config.TotalClusters; i++)
        {
            if (_fatManager.GetFatEntry(fat, i) == 0xFF) free++;
        }

        return new DiskFileSystemInfo(
            "N88-BASIC",
            (long)_config.TotalClusters * _config.ClusterSize,
            (long)free * _config.ClusterSize,
            _config.ClusterSize,
            _config.ReservedClusters * (_config.ClusterSize / _config.SectorSize),
            "PC88",
            "PC88",
            6,
            3
        );
    }

    public IEnumerable<FileEntry> GetFiles()
    {
        var files = new List<FileEntry>();
        for (int s = 0; s < _config.DirectorySectors; s++)
        {
            var dirData = ReadDirectorySector(s);
            for (int offset = 0; offset < _config.SectorSize; offset += 16)
            {
                byte mode = dirData[offset];
                if (mode == 0xFF) return files;
                if (mode == 0x00) continue;

                var entryData = new byte[16];
                Array.Copy(dirData, offset, entryData, 0, 16);
                var entry = _dirParser.Parse(entryData);

                // Enrich size from FAT chain
                var clusters = _fatManager.GetClusterChain(entry.StartCluster);
                long size = clusters.Count * _config.ClusterSize;
                
                // If it's the last cluster, check for partial sector usage if N88-BASIC supports it in FAT
                var fat = _fatManager.ReadFat();
                if (clusters.Count > 0)
                {
                    byte lastEntry = _fatManager.GetFatEntry(fat, clusters[^1]);
                    if (lastEntry >= 0xC0 && lastEntry <= 0xCF)
                    {
                        int usedSectors = lastEntry - 0xC0;
                        size = (clusters.Count - 1) * _config.ClusterSize + (usedSectors * _config.SectorSize);
                    }
                }
                
                entry = entry with { Size = size };
                files.Add(entry);
            }
        }
        return files;
    }

    public bool FileExists(string fileName)
    {
        return GetFiles().Any(f => f.FullName.Equals(fileName, StringComparison.OrdinalIgnoreCase));
    }

    public byte[] ReadFile(string fileName)
    {
        var file = GetFiles().FirstOrDefault(f => f.FullName.Equals(fileName, StringComparison.OrdinalIgnoreCase));
        if (file == null) throw new FileSystemException($"File not found: {fileName}");

        var clusters = _fatManager.GetClusterChain(file.StartCluster);
        using var ms = new MemoryStream();
        foreach (var c in clusters)
        {
            ms.Write(ReadCluster(c));
        }

        var data = ms.ToArray();
        // Trim to exact size calculated from FAT
        if (data.Length > file.Size) data = data.Take((int)file.Size).ToArray();

        if (file.Attributes.IsAscii)
        {
            return ExtractAscii(data);
        }
        return data;
    }

    public void WriteFile(string fileName, byte[] data, ExtendedFileAttributes attributes, ushort? loadAddress = null, ushort? executionAddress = null)
    {
        if (_diskContainer.IsReadOnly) throw new InvalidOperationException("Disk is read-only");
        if (FileExists(fileName)) throw new FileSystemException($"File already exists: {fileName}");

        // N88-BASIC ASCII files often end with 0x1A
        if (attributes.IsAscii && (data.Length == 0 || data[^1] != 0x1A))
        {
            var newData = new byte[data.Length + 1];
            Array.Copy(data, newData, data.Length);
            newData[^1] = 0x1A;
            data = newData;
        }

        int clustersNeeded = (data.Length + _config.ClusterSize - 1) / _config.ClusterSize;
        if (clustersNeeded == 0) clustersNeeded = 1;

        var allocatedClusters = AllocateClusters(clustersNeeded);
        if (allocatedClusters.Count < clustersNeeded) throw new FileSystemException("Not enough free space");

        WriteDataToClusters(data, allocatedClusters);

        // Update FAT
        var fat = _fatManager.ReadFat();
        for (int i = 0; i < allocatedClusters.Count; i++)
        {
            int next;
            if (i == allocatedClusters.Count - 1)
            {
                // EOF marker: C0h + sectors used in last cluster
                int lastClusterSize = data.Length % _config.ClusterSize;
                if (lastClusterSize == 0) lastClusterSize = _config.ClusterSize;
                int sectors = (lastClusterSize + _config.SectorSize - 1) / _config.SectorSize;
                next = 0xC0 + sectors;
            }
            else
            {
                next = allocatedClusters[i + 1];
            }
            _fatManager.SetFatEntry(fat, allocatedClusters[i], next);
        }
        _fatManager.WriteFat(fat);

        // Add dir entry
        var (name, ext) = ParseFileName(fileName);
        var entry = new FileEntry(name, ext, data.Length, null, DateTime.Now, attributes, allocatedClusters[0], null, null, null);
        AddDirectoryEntry(entry);
    }

    public void DeleteFile(string fileName)
    {
        if (_diskContainer.IsReadOnly) throw new InvalidOperationException("Disk is read-only");
        var file = GetFiles().FirstOrDefault(f => f.FullName.Equals(fileName, StringComparison.OrdinalIgnoreCase));
        if (file == null) throw new FileSystemException($"File not found: {fileName}");

        var clusters = _fatManager.GetClusterChain(file.StartCluster);
        var fat = _fatManager.ReadFat();
        foreach (var c in clusters) _fatManager.SetFatEntry(fat, c, 0xFF); // Mark as free
        _fatManager.WriteFat(fat);

        MarkAsDeleted(fileName);
    }

    public void RenameFile(string oldName, string newName)
    {
        if (_diskContainer.IsReadOnly) throw new InvalidOperationException("Disk is read-only");
        UpdateDirEntry(oldName, entry => {
            var (name, ext) = ParseFileName(newName);
            return entry with { FileName = name, Extension = ext };
        });
    }

    public void CopyFile(string sourceName, string targetName)
    {
        var data = ReadFile(sourceName);
        var sourceFile = GetFiles().First(f => f.FullName.Equals(sourceName, StringComparison.OrdinalIgnoreCase));
        WriteFile(targetName, data, sourceFile.Attributes);
    }

    public void UpdateAttributes(string fileName, ExtendedFileAttributes attributes)
    {
        if (_diskContainer.IsReadOnly) throw new InvalidOperationException("Disk is read-only");
        UpdateDirEntry(fileName, entry => entry with { Attributes = attributes });
    }

    public ExtendedFileAttributes CreateDefaultAttributes(bool isAscii)
    {
        // bit 0: Binary, bit 7: Tokenized (0 for ASCII)
        byte rawValue = (byte)(isAscii ? 0x00 : 0x01); 
        return new ExtendedFileAttributes(DomainAttr.None, rawValue, isAscii, "PC88");
    }

    public void Format()
    {
        if (_diskContainer.IsReadOnly) throw new InvalidOperationException("Disk is read-only");
        
        // FAT: all 0xFF
        var fatData = new byte[_config.FatSectors * _config.SectorSize];
        Array.Fill(fatData, (byte)0xFF);
        _fatManager.WriteFat(fatData);

        // Directory: all 0xFF
        var emptySector = new byte[_config.SectorSize];
        Array.Fill(emptySector, (byte)0xFF);
        for (int s = 0; s < _config.DirectorySectors; s++) WriteDirectorySector(s, emptySector);
    }

    public byte[] ReadBootArea() => _diskContainer.ReadSector(0, 0, 1);

    public void WriteBootArea(byte[] data)
    {
        if (_diskContainer.IsReadOnly) throw new InvalidOperationException("Disk is read-only");
        var sectorData = new byte[_config.SectorSize];
        Array.Copy(data, sectorData, Math.Min(data.Length, _config.SectorSize));
        _diskContainer.WriteSector(0, 0, 1, sectorData);
    }

    public FileSystemCapabilities Capabilities => 
        FileSystemCapabilities.SupportsAttributes | 
        FileSystemCapabilities.SupportsRename | 
        FileSystemCapabilities.FixedFileNameLength;

    private byte[] ReadDirectorySector(int sectorIndex)
    {
        var (track, side, sector) = GetPhysicalAddress(_config.SystemTrack, _config.DirectorySector + sectorIndex);
        return _diskContainer.ReadSector(track, side, sector);
    }

    private void WriteDirectorySector(int sectorIndex, byte[] data)
    {
        var (track, side, sector) = GetPhysicalAddress(_config.SystemTrack, _config.DirectorySector + sectorIndex);
        _diskContainer.WriteSector(track, side, sector, data);
    }

    private byte[] ReadCluster(int cluster)
    {
        int sectorsPerCluster = _config.ClusterSize / _config.SectorSize;
        var data = new byte[_config.ClusterSize];
        // In 2D, each cluster is 8 sectors (half track).
        // Cluster 0 = T0/H0/S1-8, Cluster 1 = T0/H0/S9-16, Cluster 2 = T0/H1/S1-8 ...
        int totalSectorIndex = cluster * sectorsPerCluster;
        for (int i = 0; i < sectorsPerCluster; i++)
        {
            int lba = totalSectorIndex + i;
            int track = lba / _config.SectorsPerTrack;
            int sector = (lba % _config.SectorsPerTrack) + 1;
            var (c, h, s) = GetPhysicalAddress(track, sector);
            var sectorData = _diskContainer.ReadSector(c, h, s);
            Array.Copy(sectorData, 0, data, i * _config.SectorSize, _config.SectorSize);
        }
        return data;
    }

    private void WriteCluster(int cluster, byte[] data)
    {
        int sectorsPerCluster = _config.ClusterSize / _config.SectorSize;
        int totalSectorIndex = cluster * sectorsPerCluster;
        for (int i = 0; i < sectorsPerCluster; i++)
        {
            int lba = totalSectorIndex + i;
            int track = lba / _config.SectorsPerTrack;
            int sector = (lba % _config.SectorsPerTrack) + 1;
            var (c, h, s) = GetPhysicalAddress(track, sector);
            var sectorData = new byte[_config.SectorSize];
            Array.Copy(data, i * _config.SectorSize, sectorData, 0, _config.SectorSize);
            _diskContainer.WriteSector(c, h, s, sectorData);
        }
    }

    private (int c, int h, int s) GetPhysicalAddress(int track, int sector)
    {
        if (_diskContainer.DiskType == DiskType.TwoHD)
        {
             // 2HD: 8 sectors/track, 1024 bytes/sector is common but D88 abstracts it.
             // Standard D88 parsing uses linear tracks.
             return (track / 2, track % 2, sector);
        }
        return (track / 2, track % 2, sector);
    }

    private List<int> AllocateClusters(int count)
    {
        var fat = _fatManager.ReadFat();
        var allocated = new List<int>();
        for (int i = 0; i < _config.TotalClusters && allocated.Count < count; i++)
        {
            // Skip system track (clusters overlapping T18/H1 or T40/H0)
            if (IsSystemCluster(i)) continue;
            if (_fatManager.GetFatEntry(fat, i) == 0xFF) allocated.Add(i);
        }
        return allocated;
    }

    private bool IsSystemCluster(int cluster)
    {
        int sectorsPerCluster = _config.ClusterSize / _config.SectorSize;
        int startLba = cluster * sectorsPerCluster;
        int endLba = startLba + sectorsPerCluster - 1;
        int systemLbaStart = _config.SystemTrack * _config.SectorsPerTrack;
        int systemLbaEnd = systemLbaStart + _config.SectorsPerTrack - 1;
        return (startLba <= systemLbaEnd && endLba >= systemLbaStart);
    }

    private void WriteDataToClusters(byte[] data, List<int> clusters)
    {
        int offset = 0;
        foreach (var c in clusters)
        {
            var clusterData = new byte[_config.ClusterSize];
            int toCopy = Math.Min(data.Length - offset, _config.ClusterSize);
            Array.Copy(data, offset, clusterData, 0, toCopy);
            WriteCluster(c, clusterData);
            offset += toCopy;
        }
    }

    private void AddDirectoryEntry(FileEntry entry)
    {
        for (int s = 0; s < _config.DirectorySectors; s++)
        {
            var dirData = ReadDirectorySector(s);
            for (int offset = 0; offset < _config.SectorSize; offset += 16)
            {
                byte mode = dirData[offset];
                if (mode == 0x00 || mode == 0xFF)
                {
                    _dirParser.WriteToBuffer(dirData, offset, entry);
                    WriteDirectorySector(s, dirData);
                    return;
                }
            }
        }
        throw new FileSystemException("Directory is full");
    }

    private void MarkAsDeleted(string fileName)
    {
        UpdateDirEntry(fileName, entry => null); // null means mark as 0x00
    }

    private void UpdateDirEntry(string fileName, Func<FileEntry, FileEntry?> updateFunc)
    {
        for (int s = 0; s < _config.DirectorySectors; s++)
        {
            var dirData = ReadDirectorySector(s);
            bool modified = false;
            for (int offset = 0; offset < _config.SectorSize; offset += 16)
            {
                byte mode = dirData[offset];
                if (mode == 0xFF) break;
                if (mode == 0x00) continue;
                
                var entryData = new byte[16];
                Array.Copy(dirData, offset, entryData, 0, 16);
                var entry = _dirParser.Parse(entryData);
                if (entry.FullName.Equals(fileName, StringComparison.OrdinalIgnoreCase))
                {
                    var updated = updateFunc(entry);
                    if (updated == null) dirData[offset] = 0x00;
                    else _dirParser.WriteToBuffer(dirData, offset, updated);
                    modified = true;
                    break;
                }
            }
            if (modified)
            {
                WriteDirectorySector(s, dirData);
                return;
            }
        }
        throw new FileSystemException($"File not found: {fileName}");
    }

    private (string name, string ext) ParseFileName(string fileName)
    {
        var parts = fileName.Split('.');
        string name = parts[0].ToUpper();
        if (name.Length > 6) name = name.Substring(0, 6);
        string ext = parts.Length > 1 ? parts[1].ToUpper() : "";
        if (ext.Length > 3) ext = ext.Substring(0, 3);
        return (name, ext);
    }

    private byte[] ExtractAscii(byte[] data)
    {
        var res = new List<byte>();
        foreach (var b in data)
        {
            if (b == 0x1A) break;
            res.Add(b);
        }
        return res.ToArray();
    }

    public void Dispose() { }
}
